### PR TITLE
GEODE-6798: Refactoring of client function execution logic

### PR DESCRIPTION
**Pull request by albertogpz**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by albertogpz** [comment]
Hi,
I just saw that a new pull request merged has created conflicts with my PR.
Please, do not review this PR until I fix them.

**Comment by albertogpz** [comment]
Pull request updated. Changes ready to be reviewed.

**Comment by pivotal-jbarrett** [comment]
Heads up that this PR will create conflicts. https://github.com/apache/geode/pull/3703

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 1]
```diff
@@ -121,78 +104,48 @@ public static void execute(ExecutablePool pool, String region, Function function
...
-      byte hasResult, int mRetryAttempts, boolean isHA, boolean optimizeForWrite) {
+      byte hasResult, int maxRetryAttempts) {
 
     AbstractOp op = new ExecuteRegionFunctionOpImpl(region, function, serverRegionExecutor,
-        resultCollector, hasResult, new HashSet<String>(), isHA, optimizeForWrite, true);
-    boolean reexecute = false;
-    boolean reexecuteForServ = false;
-    Set<String> failedNodes = new HashSet<String>();
-    AbstractOp reexecOp = null;
+        resultCollector, hasResult, new HashSet<String>());
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: We should avoid allocating containers if we aren't going to use them. In a stable run this method will never use the `failedNodes` parameter. Perhaps passing `null` and then checking for `null` and allocating if and only if we need it.

Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: I assume this refactor to pass in the `failedNodes` was to check it in some tests but I don't see those tests in this PR, so what is the purpose?

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 2]
```diff
@@ -310,29 +198,22 @@ public static void reexecute(ExecutablePool pool, String region, String function
 
     private FunctionException functionException;
 
-
-    public ExecuteRegionFunctionOpImpl(String region, Function function,
-        ServerRegionFunctionExecutor serverRegionExecutor, ResultCollector rc, byte hasResult,
+    private static int getMsgParts(int filterSize,
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: getMsgPartCount would be a bit clearer

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 3]
```diff
@@ -310,29 +198,22 @@ public static void reexecute(ExecutablePool pool, String region, String function
...
         Set<String> removedNodes) {
-      super(MessageType.EXECUTE_REGION_FUNCTION,
-          8 + serverRegionExecutor.getFilter().size() + removedNodes.size());
+      return 8 + filterSize + removedNodes.size();
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: add at private static final int PART_COUNT = 8;

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 4]
```diff
@@ -310,29 +198,22 @@ public static void reexecute(ExecutablePool pool, String region, String function
...
+    }
+
+    private void fillMessage(String region, Object function,
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: This is just a comment but I wanted you to consider not making the "function" parameter's type "Object". I think I see why you did it (one caller exists that just gave us a String as the function).
I'd prefer that fillMessage have two parameters: one typed as Function and the other String (call it functionId). This allows the call of serverRegionExecutor.isFnSerializationReqd() to still be here instead of in multiple callers.
I think you would just need to also test for "function" being not-null:
if (function != null && serverRegionExecutor.isFnSerializationReqd()) {
  getMessage().addStringOrObjPart(function);
} else {
  getMessage().addStringOrObjPart(functionId);
}
Other refactorings than this might also exist. I didn't like losing the type information and I didn't like the test of isFnSerializationReqd being in multiple places.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 1 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: What about Collections.emptyMap()? Logically and from jit compilation perspective it is better to have single path of execution.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 2 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: I think getMessagePartCount is even more clear. Acronyms and abbreviations should really be avoided. The IDE saves us from having to write longer method names and in reality code becomes more readable with clearer method names

Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: I agree. I will leave getMessagePartCount.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 1 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: The problem with using null instead of allocating a container with no elements is that the logic would have to be complicated in several places. Whenever we need to clear, add or remove elements to the collection we would have to first check if it is either null or empty and if so, allocate it first.
The same would happen with Collections.emptyMap() given that it returns an immutable object.
In the original implementation, the execute methods already allocated an empty HashSet at the begining of the execute methods.
I'd rather leave this part as it is in order not to embroil the logic.

Regarding the passing of failedNodes to the method, the idea was to have one single private method containing the logic of the current execute and reexecute public methods that was basically repeated. The logic is now in one method but in the case of the execute public method, the value of the failedNodes argument to pass to the new method would be an empty set.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 3 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: ok

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 4 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: I agree with you. Thanks.

**Comment by albertogpz** [comment]
> @albertogpz Thanks for working on the Geode and submitting this PR. Please address the review comments.
> I had started working on this; I thought I had updated the assignee on this ticket, but looks like I missed it. I also have additional changes which collapses execute/reexecute methods to one function. Once you push this changes, I can add that over this.
> Also, the ticket mentions about other function execution semantics (onServer, onServers), If you haven't worked on this, I have changes (work in-progress) for it and can add after your changes are merged.

@agingade Sorry if I stepped over some work you had already started. I did not work on the onServer and onServers refactoring so I will leave that to you if you have already some changes in progress.
I will address the review comments and then you can build on top of them.

**Comment by albertogpz** [comment]
> To be confident that this refactoring does not harm we need the ExecuteRegionFunctionOp to have unit test coverage of these methods. Write the test without these changes and then see if it still passes afterwards.
> The unit test should verify all the parts.
> It should also verify that the retry logic has not changed.

I assumed that the tests Bill added in ExecuteRegionFunctionOpRetryTest would be sufficient.
Is there anything else you are missing?

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 1 above)
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: Understood, and in non-critical paths I wouldn't even raise this issue. My concern is that in 99.9% of the times we execute this operation the map will be created fo no reason. When we get to optimizing function execution this map will likely be source of transient garbage production that puts pressure on the GC as a result of premature allocation. We will likely refactor it for lazy initialization then. It would be great if you could work under that assumption now when already refactoring this code.

**Comment by dschneider-pivotal** [comment]
I did not know about: ExecuteRegionFunctionOpRetryTest. I was looking at ExecuteRegionFunctionOpTest and didn't see that it covered any of the methods being refactored.
I took a quick look at ExecuteRegionFunctionOpRetryTest and it does seem like it covers the retry logic.
But I did not see it doing any verification of the Message that these methods build up. The code being modified adds a bunch of Parts to the Message and I think we want confirmation that this refactoring did not change those Parts. It looks like Anil an Bill have been in this code recently so I will defer to them on what unit test coverage is needed.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 5]
```diff
@@ -516,22 +379,23 @@ protected Object processResponse(Message msg) throws Exception {
...
-                      .getFunctionStats(functionId, executor.getRegion().getSystem())
+                      .getFunctionStats(this.functionId, this.executor.getRegion().getSystem())
                       .incResultsReceived();
+                  continue;
                 } else if (((FunctionException) result)
                     .getCause() instanceof InternalFunctionInvocationTargetException) {
                   InternalFunctionInvocationTargetException ifite =
                       (InternalFunctionInvocationTargetException) ex.getCause();
-                  failedNodes.addAll(ifite.getFailedNodeSet());
+                  this.failedNodes.addAll(ifite.getFailedNodeSet());
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: Please don't revert all the static analyzer cleanup.

Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: Sorry about that. I just sent another commit with those changes back.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 6]
```diff
@@ -63,48 +63,29 @@ private ExecuteRegionFunctionOp() {
...
+    if (!isHA) {
       maxRetryAttempts = 0;
     }
 
     do {
       try {
-        if (reexecuteForServ) {
-          reexecOp = new ExecuteRegionFunctionOpImpl(op,
+        if (isReexecute) {
+          op = new ExecuteRegionFunctionOpImpl((ExecuteRegionFunctionOpImpl) op,
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: If you change the type of `op` to `ExecuteRegionFunctionOpImpl` then this blind cast can go away. Since it’s the only type that can be passed into this method without causing a cast exception it should just be the parameter type.

Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: Thanks

**Comment by albertogpz** [comment]
> The issues I had requested have been resolved. I will defer to Bill and Anil on what unit testing is needed.
> 
> @dschneider-pivotal @albertogpz We added tests to cover the execute/reexecute path; both using unit and dunit test. This PR has additional changes w.r.t message creation, and new functions related to that; it will be nice to have unit tests for those.

I've added a new unit test class (ExecuteRegionFunctionOpImplTest) with some test cases that test the new fillMessage and getMessagePartCount methods indirectly using the constructors of ExecuteRegionFunctionOpImpl.
I am not sure if it was what was expected.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 1 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: I have sent a new commit with lazy allocation of failedNodes.
Hard to decide if the gain with less GC pressure is worth with respect to the added complexity in the code.

Comment (Bill) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: How about @kohlmu-pivotal's suggestion, @albertogpz (to use `Collections.emptyMap()`). That would address @pivotal-jbarrett's concern and keep the code simple right?

**Comment by Bill** [comment]
@dschneider-pivotal said:

> It should also verify that the retry logic has not changed.

`ExecuteRegionFunctionOpRetryTest` tests the retry logic. That's a new test that was added for the recent GEODE-6677. There are similar retry tests for the other two kinds of `Execute…Function…Op` too.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 7]
```diff
@@ -524,6 +396,7 @@ protected Object processResponse(Message msg) throws Exception {
                     .getCause() instanceof InternalFunctionInvocationTargetException) {
                   InternalFunctionInvocationTargetException ifite =
                       (InternalFunctionInvocationTargetException) ex.getCause();
+                  failedNodes = allocateFailedNodesIfEmpty(failedNodes);
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: I think this looks great!

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 1 above)
Comment (Bill) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: Woops looks like you already took @kohlmu-pivotal's advice @albertogpz 👍

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 8]
```diff
@@ -108,11 +111,19 @@ private static void execute(ExecutablePool pool, String region, String function,
...
         isReexecute = true;
         resultCollector.clearResults();
+        failedNodes = allocateFailedNodesIfEmpty(failedNodes);
         failedNodes.clear();
       }
     } while (true);
   }
 
+  private static Set<String> allocateFailedNodesIfEmpty(Set<String> failedNodes) {
+    if (failedNodes.equals(Collections.emptySet())) {
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: I feel like `equals()` is the wrong kind of equality for this condition.

When `failedNodes` is an `EmptySet`, this line calls `AbstractSet.equals()`. That method returns true because the two object references are equal.

When `failedNodes` is a `HashSet`, this line still calls `AbstractSet.equals()`. If the `HashSet` happens to be empty, then `equals()` returns true because the two objects have the same number of entries (0).

While it may be true that this method will never be called with an empty `HashSet`—there is nothing in the method itself to prevent that. We'd like to be able to look at this method in isolation and convince ourselves that it works.

I wonder if the condition would be clearer as:

```java
if (failedNodes == Collections.EMPTY_SET)
```

On the plus side that solves any problem of constructing a new `HashSet()` merely because the one referenced by `failedNodes` happens to be empty (and not a reference to `EMPTY_SET`. On the minus side it relies on `Collections.getEmptySet()` returning the same reference each time.

The minus isn't a really a minus in my mind since I believe pretty strongly that the whole point of the `Collections.getEmptySet()` method is to typecast the single (VM-wide) empty, immutable, set: `Collections.EMPTY_SET`.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 9]
```diff
@@ -108,11 +111,19 @@ private static void execute(ExecutablePool pool, String region, String function,
...
+      return new HashSet();
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: ooh and this needs to be `new HashSet<>()` to eliminate the unchecked assignment warning

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java [hunk 10]
```diff
@@ -108,11 +111,19 @@ private static void execute(ExecutablePool pool, String region, String function,
 
         isReexecute = true;
         resultCollector.clearResults();
+        failedNodes = allocateFailedNodesIfEmpty(failedNodes);
         failedNodes.clear();
       }
     } while (true);
   }
 
+  private static Set<String> allocateFailedNodesIfEmpty(Set<String> failedNodes) {
```
Comment (Bill) [inline geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: While it's true that this method allocates a new object sometimes, it seems to me that what it is really doing is ensuring that the set that it returns is mutable. Maybe it would be clearer to make the signature something like `Set<String> ensureMutability(final Set<String> failedNodes)`?

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 8 above)
Comment (pivotal-jbarrett) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: Swap the order of the equals check should address this.

Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: @Bill Good point. I agree with your proposal to solve the issue.
@pivotal-jbarrett Swapping the order of the equals in the condition would not help. It would still return true when comparing against a zero entries `HashSet`.

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 9 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: Yep. Thanks!

File: geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java (hunk 10 above)
Comment (albertogpz) [reply geode-core/src/main/java/org/apache/geode/cache/client/internal/ExecuteRegionFunctionOp.java]: @Bill I agree with you. Even if the proposed name is not perfect as you could still pass an immutable non-empty Set and the method will not convert it to a mutable Set, I think it is better than the original and good enough.
Thanks!

**Comment by agingade** [comment]
@albertogpz Do you have merge permission; can you merge your changes. If you don't have the permission, i can merge these changes.

**Comment by albertogpz** [comment]
> @albertogpz Do you have merge permission; can you merge your changes. If you don't have the permission, i can merge these changes.

I don't, so you can merge the changes.
Thanks.
